File: 114.Flatten-Binary-Tree-to-Linked-List.go
// https://leetcode.com/problems/flatten-binary-tree-to-linked-list/description/
//
// algorithms
// Medium (39.54%)
// Total Accepted:    200.6K
// Total Submissions: 507.3K

package leetcode

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func flatten(root *TreeNode) {
	resursive(root)
	return
}

func resursive(root *TreeNode) *TreeNode {
	if root == nil {
		return root
	}

	var left, right *TreeNode

	if root.Left != nil {
		left = resursive(root.Left)
	}
	if root.Right != nil {
		right = resursive(root.Right)
	}

	tmp := root

	if left != nil {
		tmp.Right = left
		tmp.Left = nil
		for left.Right != nil {
			left = left.Right
		}
		tmp = left
	}

	if right != nil {
		tmp.Right = right
	}

	return root
}

File: 114.Flatten-Binary-Tree-to-Linked-List.py
# https://leetcode.com/problems/flatten-binary-tree-to-linked-list/
#
# algorithms
# Medium (43.7%)
# Total Accepted:    256,628
# Total Submissions: 587,307


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: None Do not return anything, modify root in-place instead.
        """
        if not root:
            return None

        res = root

        while root:
            if root.left:
                pre = root.left
                while pre.right:
                    pre = pre.right
                pre.right = root.right
                root.right = root.left
                root.left = None

            root = root.right

        return res
